mobile/src/test/java/org/openhab/habdroid/model/OpenHABWidgetTest.java

```diff
@@ -0,0 +1,412 @@
+        OpenHABWidget.parseXml(sutXml, null, createXmlNode());
+        OpenHABWidget.parseJson(sut1, null, createJSONObject(1), "PNG");
+        OpenHABWidget.parseJson(sut2, null, createJSONObject(2), "SVG");
+        OpenHABWidget.parseJson(sut3, null, createJSONObject(3), "SVG");
+    }
+
+    @Test
+    public void testCountInstances() throws Exception {
+        assertEquals(sutXml.size(), 2);
+        assertEquals(sut1.size(), 2);
+        assertEquals(sut2.size(), 1);
+        assertEquals(sut3.size(), 4);
+    }
+
+    @Test
+    public void getIconPath_iconExists_returnIconUrlfromImages() throws Exception {
+        assertEquals("images/groupicon.png", sutXml.get(0).iconPath());
+    }
+
+    @Test
+    public void testGetIconPath() throws Exception {
+        assertEquals("icon/groupicon?state=OFF&format=PNG", sut1.get(0).iconPath());
+        assertEquals("icon/groupicon?state=ON&format=SVG", sut2.get(0).iconPath());
+        assertEquals("icon/slider?state=81&format=SVG", sut3.get(1).iconPath());
+        assertEquals("Rollersutter icon must always be 0 to 100, not ON/OFF", "icon/rollershutter?state=0&format=SVG", sut3.get(2).iconPath());
+        assertEquals("icon/rollershutter?state=42&format=SVG", sut3.get(3).iconPath());
+    }
+
+    @Test
+    public void testGetChildren() throws Exception {
+        assertEquals("demo11", sut1.get(1).id());
+        assertEquals("0202_0_0_1", sut3.get(2).id());
+    }
+
+    @Test
+    public void testGetPeriod() throws Exception {
+        assertEquals("M", sut1.get(0).period());
+        assertEquals("Object has no period, should default to 'D'", "D", sut2.get(0).period());
+    }
+
+    @Test
+    public void testGetStep() throws Exception {
+        assertEquals(1F, sut1.get(0).step());
+        // this is invalid in JSON (< 0), expected to be adjusted
+        assertEquals(0.1F, sut2.get(0).step());
+    }
+
+    @Test
+    public void testGetRefresh() throws Exception {
+        assertEquals(1000, sut1.get(0).refresh());
+        assertEquals("Min refresh is 100, object has set refresh to 10", 100, sut2.get(0).refresh());
+    }
+
+    @Test
+    public void testHasItem() throws Exception {
+        assertNotNull(sut1.get(0).item());
+        assertNotNull(sut2.get(0).item());
+    }
+
+    @Test
+    public void testHasLinkedPage() throws Exception {
+        assertNotNull(sut1.get(0).linkedPage());
+        assertNull(sut2.get(0).linkedPage());
+    }
+
+    @Test
+    public void testHasMappings() throws Exception {
+        assertEquals(true, sut1.get(0).hasMappings());
+        assertEquals(true, sut2.get(0).hasMappings());
+    }
+
+    @Test
+    public void testGetColors() throws Exception {
+        Integer i = new Integer(0);
+        assertEquals(i, sut1.get(0).labelColor());
+        assertEquals(i, sut1.get(0).iconColor());
+        assertEquals(i, sut1.get(0).valueColor());
+        assertEquals(i, sut2.get(0).labelColor());
```

Comment:
This test is still wrong. It passes, but has the wrong assertions. The colors if sut2 are orange, blue and red.

Reply:
That's right, but it always has been like that. We either can fix this now by copying the implementation of Color.parseColor into a mocked implementation, or we could wait for #637 after which we could assert against the color names. I'd prefer the latter ;-)

Reply:
Yes, the latter one sounds better.